pulp_container/app/viewsets.py

```diff
@@ -595,24 +596,44 @@ def add(self, request, pk):
         """
         Queues a task that creates a new RepositoryVersion by adding content units.
         """
-        add_content_units = []
+        add_content_units = {}
         repository = self.get_object()
         serializer = serializers.RecursiveManageSerializer(data=request.data)
         serializer.is_valid(raise_exception=True)
 
         if "content_units" in request.data:
             for url in request.data["content_units"]:
-                content = NamedModelViewSet.get_resource(url, Content)
-                content.touch()
-                add_content_units.append(str(content.pk))
+                add_content_units[NamedModelViewSet.extract_pk(url)] = url
+
+            self.touch_content_units(add_content_units)
 
         result = dispatch(
             tasks.recursive_add_content,
-            exclusive_resources=[repository],
-            kwargs={"repository_pk": str(repository.pk), "content_units": add_content_units},
+            [repository],
+            kwargs={
+                "repository_pk": str(repository.pk),
+                "content_units": list(add_content_units.keys()),
+            },
         )
         return OperationPostponedResponse(result, request)
 
+    def touch_content_units(self, content_units):
+        """Validate and touch referenced content units."""
+        content_units_pks = content_units.keys()
+        existing_content_units_pks = Content.objects.filter(
+            pk__in=content_units.keys()
+        ).values_list("pk", flat=True)
+        existing_content_units_pks = {str(pk) for pk in existing_content_units_pks}
+
+        missing_pks = content_units_pks - existing_content_units_pks
+        if missing_pks:
+            missing_hrefs = [content_units[pk] for pk in missing_pks]
+            raise ValidationError(
+                _("Could not find the following content units: {}").format(missing_hrefs)
+            )
+
+        Content.objects.filter(pk__in=content_units.keys()).touch()
```

Comment:
Please make it first touch then verify.
Also keeping the query in a variable may help. Like: `qs = Content.objects.filter(pk__in=content_unit_pks)`.

Reply:
Could you please provide any specific reason why should I touch content before verifying it?

Reply:
because touching prevents it from disappearing. If you first check, and then touch, you leave a window open where the content is gone by the time you attempt to touch it.